volatility3/framework/objects/utility.py

```diff
@@ -94,16 +99,113 @@ def pointer_to_string(
         count=count,
         errors=errors,
         block_size=block_size,
+        encoding=encoding,
     )
 
 
+def gather_contiguous_bytes_from_address(
+    context, data_layer, starting_address: int, count: int
+) -> bytes:
+    """
+    This method reconstructs a string from memory while also carefully examining each page
+
+    It goes page-by-page reading the bytes. This is done by calculating page boundaries
+    and then only reading one page at a time.
+
+    If a page is missing, the code initially catches the exception.
+    If data is non-empty (meaning at least one read succeeded), then we return what was read
+    If the first page fails, then we re-raise the exception
+    """
+
+    data = b""
+
+    if isinstance(data_layer, interfaces.layers.TranslationLayerInterface):
+        last_address = None
+
+        for address, length, _, _, _ in data_layer.mapping(
+            offset=starting_address, length=count, ignore_errors=True
+        ):
+            # Used to track when we hit a paged out page
+            if not last_address:
+                last_address = address + length
+
+            # we hit a swapped out page
+            elif last_address and last_address != address:
+                break
+
+            data += data_layer.read(address, length)
+
+            last_address = address + length
+
+    elif starting_address + count < data_layer.maximum_address:
+        data = data_layer.read(starting_address, count)
+
+    # if we were able to read from the first page, we want to try and construct the string
```

Comment:
I think there's *technically* the case where the starting address, doesn't being where the search is supposed to...

```
|   |<---asked for--->|  |
|paged out|actual data|   |
```

Reply:
Hmm you will have to explain more. If we got a later page first (actual data in your diagram), then the `last_address` check would fail and we wouldn't read.

Reply:
Yeah, but before we set `last_address` at the start of the loop based on the first entry (where `address` may not have been `starting_address`).  I think we're all good now.  5:)